### PR TITLE
Action: check topic frequency

### PR DESCRIPTION
**Pull request by Nikhil-Singhal-06**
### Description
New action for checking specific topic frequency

### Pull request checklist

- [ ] [Issue](https://github.com/IntelLabs/scenario_execution/issues) created that explains the change and why it's needed
- [x] Tests are part of the PR (for bug fixes / features)
- [x] [Docs](https://intellabs.github.io/scenario_execution/) reviewed and added / updated if needed (for bug fixes / features)
- [x] Format and Lint checks (`make check`) pass locally
- [x] PR applies Apache 2.0 License to all code files

### Pull request type

- [x] Feature
- [x] Documentation changes

### Current behavior
No action available to check topic frequency

### New behavior
new action assert_topic_latency

### CODE REVIEW

File: docs/libraries.rst [hunk 1]
```diff
@@ -180,6 +180,18 @@ Wait until a defined distance was traveled, based on odometry.
 - ``namespace: string``:  Namespace of the odometry topic
 - ``distance: length``: Traveled distance at which the action succeeds.
 
+``assert_topic_latency()``
+"""""""""""""""""""""""""""""""""""""""""""""""""""""""""
+
+check for topic latency.
```
Comment (fmirus) [inline docs/libraries.rst:186]: Imho, this description could be a little more detailed. Particularly, what this action is intended for.

File: scenario_execution_gazebo/scenario_execution_gazebo/actions/gazebo_wait_for_sim.py [hunk 2]
```diff
@@ -42,6 +42,8 @@ def __init__(self, name, world_name: str, **kwargs):
                    world_name + "/clock", "-e", "--json-output", "-n", "1"]
         super().__init__(name, command)
         self.current_state = WaitForSimulationActionState.IDLE
+        self.start_time = time.time()
+        self.timeout_sec = 60
```
Comment (fmirus) [inline scenario_execution_gazebo/scenario_execution_gazebo/actions/gazebo_wait_for_sim.py:46]: why a hard-coded timeout?

File: scenario_execution_ros/scenario_execution_ros/lib_osc/ros.osc [hunk 3]
```diff
@@ -107,3 +107,11 @@ action differential_drive_robot.tf_close_to:
 action differential_drive_robot.odometry_distance_traveled:
     namespace: string = ''
     distance: length
+
+action assert_topic_latency:
+    topic_name: string # topic name to wait for message
+    latency: time #the time to compare against
+    comparison_operator: comparison_operator = comparison_operator!le # the comparision is done using the this operator
```
Comment (fmirus) [inline scenario_execution_ros/scenario_execution_ros/lib_osc/ros.osc:114]: typo: the this

**Review by fmirus — CHANGES_REQUESTED** [review]

File: scenario_execution_ros/scenario_execution_ros/actions/assert_topic_latency.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+                    result = py_trees.common.Status.FAILURE
+                    self.feedback_message = f'latency of the topic {self.topic_name} exceeds the given latency.'  # pylint: disable= attribute-defined-outside-init
+            else:
+                result = py_trees.common.Status.RUNNING
+        return result
+
+    def _callback(self, msg):
+        if self.wait_for_first_message:
+            self.wait_for_first_message = False
+        else:
```
Comment (fmirus) [inline scenario_execution_ros/scenario_execution_ros/actions/assert_topic_latency.py:100]: isn't this else preventing the node from actually processing the very first message it receives?

**Review by fmirus — CHANGES_REQUESTED** [review]

File: scenario_execution_ros/scenario_execution_ros/actions/assert_topic_latency.py (hunk 4 above)
Comment (Nikhil-Singhal-06) [reply scenario_execution_ros/scenario_execution_ros/actions/assert_topic_latency.py:100]: Yes, but I don't think this has any effect on the behavior of the action.

File: scenario_execution_ros/test/test_assert_topic_latency.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+            msg.data = 'Hello'
+            self.publisher.publish(msg)
+
+    def tearDown(self):
+        self.running = False
+        self.publishing_thread.join()
+        self.node.destroy_node()
+        rclpy.try_shutdown()
+
+    def test_success(self):
```
Comment (fred-labs) [inline scenario_execution_ros/test/test_assert_topic_latency.py:62]: please add a failing test

**Review by fred-labs — CHANGES_REQUESTED** [review]

File: scenario_execution_ros/test/test_assert_topic_latency.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
```
Comment (fred-labs) [inline scenario_execution_ros/test/test_assert_topic_latency.py:63]: I previously missed that. This test is actually also a "failing" test, right? (only that the action exits with success).
Please also add a test that shows, that the action keeps running while a topic comes in correctly (e.g. for a few seconds).

**Review by fred-labs — CHANGES_REQUESTED** [review]

File: scenario_execution_ros/test/test_assert_topic_latency.py (hunk 6 above)
Comment (Nikhil-Singhal-06) [reply scenario_execution_ros/test/test_assert_topic_latency.py:63]: I think that we are not at the same level in terms of understanding how action should work.

First,
What exactly is the failing case?
If the latency specified in the scenario exceeds the actual latency of the topic. 

If so, then the scenario will succeed or fail depending on whether fail_on_finish is true or false.

Secondly,
the action will continue running if the actual latency is less than the specified latency regardless fail_on_finish is true or false, correct?

is this correct or am I missing something?

Comment (fred-labs) [reply scenario_execution_ros/test/test_assert_topic_latency.py:63]: That's correct. So a "good case" would be something like this:

```
scenario test_assert_topic_latency:
    do parallel:
        serial:
            assert_topic_latency() with:
                keep(it.topic_name == '/bla')
                keep(it.latency == 1.5s)
            emit fail
        serial:
            wait elapsed(10s)
            emit end
```

**Review by fred-labs — CHANGES_REQUESTED** [review]

File: scenario_execution_ros/test/test_assert_topic_latency.py (hunk 6 above)
Comment (Nikhil-Singhal-06) [reply scenario_execution_ros/test/test_assert_topic_latency.py:63]: Ok, perfect

**Pull request closed without merging**